core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -322,68 +313,67 @@ private void stop(CompletableFuture<Void> future) {
-        }
-        return workerGroup.next();
+        return config().workerGroup().next();
     }
 
     private void stop0(CompletableFuture<Void> future) {
         assert future != null;
 
-        final EventLoopGroup bossGroup = this.bossGroup;
+        final EventLoopGroup bossGroup = config.bossGroup();
         final GracefulShutdownSupport gracefulShutdownSupport = this.gracefulShutdownSupport;
 
         if (gracefulShutdownSupport == null) {
-            stop1(future, bossGroup);
+            stop1(future);
             return;
         }
 
         // Check every 100 ms for the server to have completed processing requests.
         final ScheduledFuture<?> quietPeriodFuture = bossGroup.scheduleAtFixedRate(() -> {
             if (gracefulShutdownSupport.completedQuietPeriod()) {
-                stop1(future, bossGroup);
+                stop1(future);
             }
         }, 0, 100, TimeUnit.MILLISECONDS);
 
         // Make sure the event loop stops after the timeout, regardless of what
         // the GracefulShutdownSupport says.
         bossGroup.schedule(() -> {
             quietPeriodFuture.cancel(false);
-            stop1(future, bossGroup);
+            stop1(future);
         }, config.gracefulShutdownTimeout().toMillis(), TimeUnit.MILLISECONDS);
     }
 
-    private void stop1(CompletableFuture<Void> future, EventLoopGroup bossGroup) {
-        // FIXME(trustin): Shutdown and terminate the blockingTaskExecutor.
-        //                 Could be fixed while fixing https://github.com/line/armeria/issues/46
-
-        final Future<?> bossShutdownFuture;
-        if (bossGroup != null) {
-            bossShutdownFuture = bossGroup.shutdownGracefully();
-            this.bossGroup = null;
-        } else {
-            bossShutdownFuture = ImmediateEventExecutor.INSTANCE.newSucceededFuture(null);
-        }
-
-        bossShutdownFuture.addListener(f1 -> {
-            // All server ports have been unbound.
-            primaryActivePort = null;
-            activePorts.clear();
-
-            // Shut down the workers.
-            final EventLoopGroup workerGroup = this.workerGroup;
-            final Future<?> workerShutdownFuture;
-            if (workerGroup != null) {
-                workerShutdownFuture = workerGroup.shutdownGracefully();
-                this.workerGroup = null;
+    private void stop1(CompletableFuture<Void> future) {
+        // Close all server sockets.
+        ChannelUtil.close(serverChannels).whenComplete((unused1, unused2) -> {
+            // Shut down the boss group if necessary.
+            final Future<?> bossShutdownFuture;
+            if (config.shutdownBossGroupOnStop()) {
+                bossShutdownFuture = config.bossGroup().shutdownGracefully();
             } else {
-                workerShutdownFuture = ImmediateEventExecutor.INSTANCE.newSucceededFuture(null);
+                bossShutdownFuture = ImmediateEventExecutor.INSTANCE.newSucceededFuture(null);
```

Comment:
Not sure why I shut down the worker group first before, but it's better shutting down the boss group first so that no one can accept an incoming connection.